src/katgpucbf/fgpu/engine.py

```diff
@@ -721,20 +721,21 @@ async def run_processing(self) -> None:
         logger.debug("run_processing completed")
         self._out_queue.put_nowait(None)
 
-    def _chunk_finished(self, chunk: send.Chunk, future: asyncio.Future) -> None:
-        """Return a chunk to the free queue after it has completed transmission.
-
-        This is intended to be used as a callback on an :class:`asyncio.Future`.
-        """
-        if chunk.cleanup is not None:
-            chunk.cleanup()
-            chunk.cleanup = None  # Potentially helps break reference cycles
+    async def _chunk_send_and_cleanup(
+        self, streams: list["spead2.send.asyncio.AsyncStream"], n_frames: int, chunk: send.Chunk
+    ) -> None:
+        """Return a chunk to the free queue after it has completed transmission."""
         try:
-            future.result()  # No result, but want the exception
+            task = asyncio.create_task(chunk.send(streams, n_frames, self.engine.time_converter, self.engine.sensors))
+            await task
```

Comment:
```suggestion
            await chunk.send(streams, n_frames, self.engine.time_converter, self.engine.sensors)
```
No need to create an asynchronous task which you immediately synchronise with.

Reply:
If you do 
```python
await coroutine()
```
then it won't get garbage-collected, because the interpreter is waiting at this point for the coroutine to finish before it continues execution of this function.

If you want to start a coroutine and carry on with some other stuff while you wait for it to finish, then something like this:
```python
my_task = create_task(coroutine())
foo = 1 + 1
bar = 2 + 2
# Now all the stuff we wanted to do is finished, we just need to wait for the coroutine
await my_task # If the task is already finished, this will just be nothing, but if we need to wait for it, execution will pause here
return foo + bar
```
this is fine, because the task has a handle and won't get garbage collected.

Now the problem comes in, say you have a task that needs to get done in the background, you don't really care when, but it just must not get forgotten.


```python
create_task(coroutine())
foo = 1 + 1
bar = 2 + 2
return foo + bar
```
now the task doesn't have a handle, the function returns, etc. so the interpreter doesn't have a way to know whether or not it still needs to be there, so is in danger of being garbage-collected. This is the kind of thing that we want to avoid. There seems to be (based on our experience) a low probability of this happening, but when you do it enough times, it's certainly going to come up now and again and the origin of the ticket is that it's going to be difficult to debug this, so rather just program defensively. So we avoid this by using the `add_service_task(my_task)` mechanism which maintains a reference to the task and prevents it being garbage collected.

If that makes sense.

Reply:
Thanks @james-smith-za, that's a good explanation.

A short rule of thumb is that any task or coroutine that you `await` is safe from this issue, because the fact that you're able to refer to it to `await` it means you still have a reference to it.